Add an option to enable Multipath TCP on clients

Aperence opened the pull request:
Multipath TCP (MPTCP) is a TCP extension allowing to enhance the reliability of the network by using multiple interfaces. This extension provides a seamless handover between interfaces in case of deterioration of the connection on the original one. In the context of iOS and Mac OS X, it could be really interesting to leverage the capabilities of MPTCP as they could benefit from their multiple interfaces (ethernet + Wi-fi for Mac OS X, Wi-fi + cellular for iOS).

This contribution introduces patches to HTTPClient.Configuration and establishment of the Bootstraps. A supplementary field "enableMultipath" was added to the configuration, allowing to request the use of MPTCP. This flag is then used when creating the channels to configure the client.

Note that in the future, it might also be potentially interesting to offer more precise configuration options for MPTCP on MacOS, as the Network framework allows also to select a type of service, instead of just offering the option to create MPTCP connections. Currently, when enabling MPTCP, only the Handover mode is used.

Lukasa commented:
Hi @Aperence, sorry for the delay in getting a review here. I'll take a look now.

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -755,22 +760,25 @@ public class HTTPClient {
                     ignoreUncleanSSLShutdown: Bool = false,
-                    decompression: Decompression = .disabled) {
+                    decompression: Decompression = .disabled,
+                    enableMultipath: Bool = false) {
```
Unfortunately, we can't add new parameters to public methods. This will need a new method that has the extra parameter.

Aperence replied:
Does it require a new method, or do you prefer simply having the field enableMPTCP as public in the configuration, allowing to set it freely ?

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -738,14 +738,19 @@ public class HTTPClient {
             ignoreUncleanSSLShutdown: Bool = false,
-            decompression: Decompression = .disabled
+            decompression: Decompression = .disabled,
+            enableMultipath: Bool = false
```
Unfortunately, we can't add new parameters here. We need a new method, with the new parameter.

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -780,7 +788,8 @@ public class HTTPClient {
                     ignoreUncleanSSLShutdown: Bool = false,
-                    decompression: Decompression = .disabled) {
+                    decompression: Decompression = .disabled,
+                    enableMultipath: Bool = false) {
```
Same note here.

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -799,7 +809,8 @@ public class HTTPClient {
                     decompression: Decompression = .disabled,
-                    backgroundActivityLogger: Logger?) {
+                    backgroundActivityLogger: Logger?,
+                    enableMultipath: Bool = false) {
```
Same note here.

Lukasa commented on Sources/AsyncHTTPClient/HTTPClient.swift:
```diff
@@ -808,23 +819,26 @@ public class HTTPClient {
                     ignoreUncleanSSLShutdown: Bool = false,
-                    decompression: Decompression = .disabled) {
+                    decompression: Decompression = .disabled,
+                    enableMultipath: Bool = false) {
```
Same note here.

Lukasa reviewed (changes requested):
Thanks for this! Generally it looks like a great addition. I've noted that we need to tweak the API surface a bit.

Can I also ask that you add at least one unit test that uses the feature? That'll ensure that we don't break it.

Aperence commented:
Hello @Lukasa,

First, thanks for the review !

I've removed the changes to the methods' signatures. It is still possible to set the enableMultipath field as it is public, but if you prefer, I can also introduce a method for it. 

I've added also another test that enables MPTCP, and check that we can still fetch https://httpbin.org/get, as requested.

I remain available if other changes are needed. 
Have a nice day !

Lukasa commented:
@swift-server-bot test this please

Lukasa commented on Tests/AsyncHTTPClientTests/HTTPClientTests.swift:
```diff
@@ -3589,6 +3589,17 @@ final class HTTPClientTests: XCTestCaseHTTPClientTestsBaseClass {
         let response = try client.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
         XCTAssertEqual(.ok, response.status)
     }
+    
```
The soundness script doesn't like the whitespace here

```suggestion

```

Lukasa commented:
Ok, looks like the linux tests fail with `EINVAL`. This happens with kernels that don't support MPTCP, so you'll need to add a try-catch that checks for this case and skips the test in that context.

Aperence commented:
Just added the do-catch now !

Lukasa commented on Tests/AsyncHTTPClientTests/HTTPClientTests.swift:
```diff
@@ -3590,6 +3590,21 @@ final class HTTPClientTests: XCTestCaseHTTPClientTestsBaseClass {
     }
 
+    func testClientWithMultipath() throws {
+        do{
```
```suggestion
        do {
```

Lukasa commented on Tests/AsyncHTTPClientTests/HTTPClientTests.swift:
```diff
@@ -3590,6 +3590,21 @@ final class HTTPClientTests: XCTestCaseHTTPClientTestsBaseClass {
+            }
+            let response = try client.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
+            XCTAssertEqual(.ok, response.status)
+        }catch{
```
We need to catch the specific error, as other errors should still fail the test. This should be [`IOError`](https://swiftpackageindex.com/apple/swift-nio/2.72.0/documentation/niocore/ioerror), with `errnoCode == EINVAL`. Please also `throw XCTSkip()` in the catch as well.

Aperence commented:
Applied required changes

Lukasa commented on Tests/AsyncHTTPClientTests/HTTPClientTests.swift:
```diff
@@ -3590,6 +3590,22 @@ final class HTTPClientTests: XCTestCaseHTTPClientTestsBaseClass {
+            }
+            let response = try client.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
+            XCTAssertEqual(.ok, response.status)
+        }catch let error as IOError where error.errnoCode == EINVAL{
```
Quick cleanup:

```suggestion
        } catch let error as IOError where error.errnoCode == EINVAL {
```

Lukasa commented:
@swift-server-bot test this please

Lukasa commented on Tests/AsyncHTTPClientTests/HTTPClientTests.swift:
```diff
@@ -3590,6 +3590,22 @@ final class HTTPClientTests: XCTestCaseHTTPClientTestsBaseClass {
+            }
+            let response = try client.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
+            XCTAssertEqual(.ok, response.status)
+        } catch let error as IOError where error.errnoCode == EINVAL {
```
Just tried running this locally, it turns out `EPROTONOSUPPORT` is also sometimes thrown on Linux, so we need to add that to the list.

Aperence replied:
Added this, I've checked the [documentation](https://www.mptcp.dev/implementation.html) and it seems that it could also raise an `ENOPROTOOPT` error, so I've added this one too to the list

Lukasa commented:
@swift-server-bot test this please

Lukasa reviewed (approved):
Nice one, thanks!
